test/e2e/storage/drivers/base.go

```diff
@@ -0,0 +1,143 @@
+type PreprovisionedVolumeTestDriver interface {
+	TestDriver
+	// CreateVolume creates a pre-provisioned volume.
+	CreateVolume(testpatterns.TestVolType)
+	// DeleteVolume deletes a volume that is created in CreateVolume
+	DeleteVolume(testpatterns.TestVolType)
+}
+
+// InlineVolumeTestDriver represents an interface for a TestDriver that supports InlineVolume
+type InlineVolumeTestDriver interface {
+	PreprovisionedVolumeTestDriver
+	// GetVolumeSource returns a volumeSource for inline volume.
+	// It will set readOnly and fsType to the volumeSource, if TestDriver supports both of them.
+	// It will return nil, if the TestDriver doesn't support either of the parameters.
+	GetVolumeSource(readOnly bool, fsType string) *v1.VolumeSource
+}
+
+// PreprovisionedPVTestDriver represents an interface for a TestDriver that supports PreprovisionedPV
+type PreprovisionedPVTestDriver interface {
+	PreprovisionedVolumeTestDriver
+	// GetPersistentVolumeSource returns a PersistentVolumeSource for pre-provisioned Persistent Volume.
+	// It will set readOnly and fsType to the PersistentVolumeSource, if TestDriver supports both of them.
+	// It will return nil, if the TestDriver doesn't support either of the parameters.
+	GetPersistentVolumeSource(readOnly bool, fsType string) *v1.PersistentVolumeSource
+}
+
+// DynamicPVTestDriver represents an interface for a TestDriver that supports DynamicPV
+type DynamicPVTestDriver interface {
+	TestDriver
+	// GetDynamicProvisionStorageClass returns a StorageClass dynamic provision Persistent Volume.
+	// It will set fsType to the StorageClass, if TestDriver supports it.
+	// It will return nil, if the TestDriver doesn't support it.
+	GetDynamicProvisionStorageClass(fsType string) *storagev1.StorageClass
+}
+
+// DriverInfo represents a combination of parameters to be used in implementation of TestDriver
+type DriverInfo struct {
+	Name       string // Name of the driver
+	FeatureTag string // FeatureTag for the driver
+
+	MaxFileSize        int64           // Max file size to be tested for this driver
+	SupportedFsType    map[string]bool // Map of string for supported fs type
+	IsPersistent       bool            // Flag to represent whether it provides persistency
+	IsFsGroupSupported bool            // Flag to represent whether it supports fsGroup
+	IsBlockSupported   bool            // Flag to represent whether it supports Block Volume
+
+	// Parameters below will be set inside test loop by using SetCommonDriverParameters.
+	// Drivers that implement TestDriver is required to set all the above parameters
+	// and return DriverInfo on GetDriverInfo() call.
+	Framework *framework.Framework       // Framework for the test
+	Config    framework.VolumeTestConfig // VolumeTestConfig for thet test
+}
+
+// GetDriverNameWithFeatureTags returns driver name with feature tags
+// For example)
+// In gce:
+//  - [Driver: nfs]]
+//  - [Driver: rbd[Feature:Volumes]]
```

Comment:
I'm not sure if nested tags will mess up the regexes that test-infra uses.  It would be better to not nest them

Reply:
Fixed as suggested.